### PR TITLE
Feature/iat 3475

### PR DESCRIPTION
**Pull request by bdrainer**
The Gateway when configured for OpenAM (not Okta) will pull user permission from the perm service.  It then encodes the user and puts it in the outbound request as the `Authorization` header.  Receiving services then decode that header shoving it in their security context.  This is almost exactly what you will see in the TIMS UI.

### CODE REVIEW

File: src/main/java/org/opentestsystem/ap/apigateway/ApiGatewayConfiguration.java [hunk 1]
```diff
@@ -59,12 +121,25 @@ public void configure(final HttpSecurity http) throws Exception {
         };
     }
 
+    @Bean
+    public PermissionServiceClient permissionServiceClient() {
+        return new PermissionServiceClient(
+            this.restTemplateBuilder,
+            this.itemBankProperties.getServices().getPermission()
+        );
+    }
```
Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/apigateway/ApiGatewayConfiguration.java]: This is the client to the permission service

File: src/main/java/org/opentestsystem/ap/apigateway/ApiGatewayConfiguration.java [hunk 2]
```diff
@@ -59,12 +121,25 @@ public void configure(final HttpSecurity http) throws Exception {
...
+
+    @Bean
+    public TimsUserDetailsService timsUserDetailsService() {
+        return new TimsUserDetailsService(permissionServiceClient());
+    }
```
Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/apigateway/ApiGatewayConfiguration.java]: This is the user detail service used by the auth process for getting the principal that is ultimately put in the security context.

File: src/main/java/org/opentestsystem/ap/apigateway/ApiGatewayConfiguration.java [hunk 3]
```diff
@@ -80,8 +155,14 @@ public void handleError(ClientHttpResponse response) throws IOException {
...
         });
 
+        // this converter ultimately sets the OAuth2Authentication.userAuthentication with
+        // an instance of type UsernamePasswordAuthenticationToken
+        DefaultUserAuthenticationConverter authenticationConverter = new DefaultUserAuthenticationConverter();
+        authenticationConverter.setUserDetailsService(timsUserDetailsService());
+
         DefaultAccessTokenConverter tokenConverter = new DefaultAccessTokenConverter();
         tokenConverter.setIncludeGrantType(true);
+        tokenConverter.setUserTokenConverter(authenticationConverter);
```
Comment (bdrainer) [inline src/main/java/org/opentestsystem/ap/apigateway/ApiGatewayConfiguration.java]: This configures the auth process to utilize our user detail service

File: src/main/resources/application.yml [hunk 4]
```diff
@@ -13,19 +13,68 @@ info:
...
+      url: "http://localhost:8182"
+
+jwt:
+  secret: "\${TIMS_JWT_SECRET}"
+
 zuul:
+  host:
+    socket-timeout-millis: 30000
   routes:
+    auth-api:
```
Comment (millerts) [inline src/main/resources/application.yml]: Are all these routes used by this service?  If not, I recommend removing those that are unused.

Comment (bdrainer) [reply src/main/resources/application.yml]: Only two of these are required.  The others are not and I will remove them however, I did find it useful.  We can longer use Postman to call any of the services directly.  Previously we could and we were treated as anonymous, but now you have to have a valid Authorization token in the header or else the app rejects the call.  

Using the API Gateway was a way to solve that.  You get a token from OpenAM then call whatever service you want via the gateway.

**Pull request merged**